File: leetcode_0669/cpp/leetcode_0669.cpp
//
// Created by wangzy on 2022/4/4.
//

#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>
#include <random>
#include <ctime>

using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0),left(nullptr),right(nullptr) {}
    TreeNode(int x) : val(x),left(nullptr),right(nullptr) {}
    TreeNode(int x,TreeNode *left,TreeNode* right):val(x),left(left),right(right) {}
};

class Solution {
public:
    TreeNode* trimBST(TreeNode* root, int low, int high) {
        //修剪节点 递归判定，如果存在当前节点在[low,high]范围内则继续遍历子树
        //如果当前节点小于low,则返回 右子树满足的节点
        //如果当前节点大于high,则返回 左子树满足的节点
        if(root){
            if(root->val >= low && root->val <= high){
                root->left = trimBST(root->left,low,high);
                root->right = trimBST(root->right,low,high);
                return root;
            }else if(root->val < low){
                return trimBST(root->right,low,high);
            }else if(root->val > high){
                return trimBST(root->left,low,high);
            }
        }
        return nullptr;
    }
    vector<string> printTree(TreeNode* root){
        string s;
        vector<string> levelOrder = levelOrderStr(root);
        return levelOrder;
    }
    void deleteTree(TreeNode* root){
        if(root){
            deleteTree(root->left);
            deleteTree(root->right);
            root = nullptr;
        }
    }
    vector<string> levelOrderStr(TreeNode* root){
        vector<TreeNode*> cur;
        vector<TreeNode*> next;
        if(!root)
            return vector<string> {};
        if(root)
            cur.push_back(root);
        vector<string> res;
//        vector<string> curEles;
        while(cur.size()){
            //标识next是否都为null
            int flag = 0;
            for(int i = 0;i < cur.size();i++){
                TreeNode* node = cur[i];
                if(!node){
                    res.push_back("null");
                    continue;
                }
                else
                    res.push_back(to_string(node->val));
                //如果node的孩子不为null,则改变flag
                if(node->left || node->right)
                    flag++;
                next.push_back(node->left);
                next.push_back(node->right);
            }
            cur = next;
            if(flag == 0)
                cur.clear();
            next.clear();
        }
        return res;
    }
};


template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

int main() {
    TreeNode* root = new TreeNode(1);
    TreeNode* left = new TreeNode(0);
    TreeNode* right = new TreeNode(2);
//    TreeNode* ll = nullptr;
//    TreeNode* lr = nullptr;
//    TreeNode* rl = new TreeNode(3);
//    TreeNode* rr = new TreeNode(6);
    root->left = left;
    root->right = right;
//    left->left = ll;
//    left->right = lr;
//    right->left = rl;
//    right->right = rr;
    vector<string> rootArr = Solution().levelOrderStr(root);
    printVec(rootArr);
    //root = [1,0,2], low = 1, high = 2
    int low = 1;
    int high = 2;
    TreeNode* anoRoot = Solution().trimBST(root,low,high);
    //[1,null,2]
    vector<string> anoArr = Solution().levelOrderStr(anoRoot);
    printVec(anoArr);
}

File: leetcode_0669/java/leetcode_0669.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class leetcode_0669 {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    class Solution {
        public TreeNode trimBST(TreeNode root, int low, int high) {
            //修剪节点 递归判定，如果存在当前节点在[low,high]范围内则继续遍历子树
            //如果当前节点小于low,则返回 右子树满足的节点
            //如果当前节点大于high,则返回 左子树满足的节点
            if(root != null){
                if(root.val >= low && root.val <= high){
                    root.left = trimBST(root.left,low,high);
                    root.right = trimBST(root.right,low,high);
                    return root;
                }else if(root.val < low)
                    return trimBST(root.right,low,high);
                else if(root.val > high)
                    return trimBST(root.left,low,high);
            }
            return null;
        }

        public List<List<Integer>> levelOrder(TreeNode root) {
            List<TreeNode> cur = new ArrayList();
            List<TreeNode> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<List<Integer>> res = new LinkedList();
            List<Integer> curEles = new LinkedList();
            while (cur.size() > 0) {
                for (int i = 0; i < cur.size(); i++) {
                    curEles.add(cur.get(i).val);
                    if (cur.get(i).left != null)
                        next.add(cur.get(i).left);
                    if (cur.get(i).right != null)
                        next.add(cur.get(i).right);
                }
                res.add(new LinkedList(curEles));
                cur = new LinkedList(next);
                curEles.clear();
                next.clear();
            }
            return res;
        }

        public List<String> levelOrderStr(TreeNode root) {
            List<TreeNode> cur = new ArrayList();
            List<TreeNode> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<String> res = new LinkedList();
            while (cur.size() > 0) {
                int flag = 0;
                for (int i = 0; i < cur.size(); i++) {
                    if (cur.get(i) == null) {
                        res.add("null");
                        continue;
                    } else
                        res.add(String.valueOf(cur.get(i).val));
                    TreeNode left = cur.get(i).left;
                    TreeNode right = cur.get(i).right;
                    if (left != null || right != null)
                        flag++;
                    next.add(left);
                    next.add(right);
                }
                cur = new LinkedList(next);
                if (flag == 0)
                    cur.clear();
                next.clear();
            }
            return res;
        }

        public <T> void printListArr(List<List<T>> list) {
            for (int i = 0; i < list.size(); i++) {
                if (i == list.size() - 1) {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.println(Arrays.toString(list.get(i).toArray(Integer[]::new)) + " ] ");
                } else {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.printf(Arrays.toString(list.get(i).toArray(Integer[]::new)) + " , ");
                }
            }
        }
    }
    public static void main(String[] args) {
        leetcode_0669.Solution solution = new leetcode_0669().new Solution();
        leetcode_0669.TreeNode root = new leetcode_0669().new TreeNode(1);
        leetcode_0669.TreeNode left = new leetcode_0669().new TreeNode(0);
        leetcode_0669.TreeNode right = new leetcode_0669().new TreeNode(2);
//        leetcode_0669.TreeNode ll = null;
//        leetcode_0669.TreeNode lr = null;
//        leetcode_0669.TreeNode rl = new leetcode_0669().new TreeNode(3);
//        leetcode_0669.TreeNode rr = new leetcode_0669().new TreeNode(6);
//        left.left = ll;
//        left.right = lr;
//        right.left = rl;
//        right.right = rr;
        root.left = left;
        root.right = right;
//      root = [1,0,2], low = 1, high = 2
        List<String> rootStr = solution.levelOrderStr(root);
        System.out.println(Arrays.toString(rootStr.toArray(String[]::new)));

        int low = 1;int high = 2;
        TreeNode anoRoot = solution.trimBST(root, low, high);
        List<String> anoStr = solution.levelOrderStr(anoRoot);
        System.out.println(Arrays.toString(anoStr.toArray(String[]::new)));

        //result
        //[1,null,2]
    }
}
